README.md
**flutter 应用升级插件**

**升级功能说明：**

	应用程序升级是 App 的基础功能之一，下面介绍应用程序升级的方式：
	从平台方面来说：
	IOS：跳转到 app store 升级。
	Android：跳转到应用市场升级，或下载 apk 包升级。
	从强制性来说：
	强制升级：用户必须升级才能继续使用 App。
	非强制升级：用户可以选择取消升级当前版本。

**升级流程说明：**

	访问后台接口获取是否有新的版本，IOS 还可通过https://itunes.apple.com/cn/lookup?bundleId=*** 来获取。
	有新版本则弹出升级提示框，根据当前版本是否是强制更新控制是否显示取消按钮。
	用户如选择取消，提示框消失。
	用户如选择升级，IOS 跳转 app store 完成更新，Android 下载 apk 完成后跳转到 apk 安装页面。

**安装说明：**

	在pubspec.yaml中加入：
		flutter_app_upgrader:
			git:
				url: https://github.com/flyaswind/flutter_app_upgrader.git
	执行flutter命令获取包
		flutter pub get
	引入
		import 'package:flutter_app_upgrader/flutter_app_upgrader.dart';

	安卓平台需要的额外配置：
		安卓清单文件./android/app/src/main/AndroidManifest.xml 需要含有以下配置
		权限配置：
			<uses-permission android:name="android.permission.INTERNET" />
			<uses-permission android:name="android.permission.REQUEST_INSTALL_PACKAGES" />

		application 结点内添加
			<provider
				android:name="androidx.core.content.FileProvider"
				android:authorities="{包名}.fileprovider"
				android:exported="false"
				android:grantUriPermissions="true">
				<meta-data
						android:name="android.support.FILE_PROVIDER_PATHS"
						android:resource="@xml/file_paths" />
			</provider>

		./android/app/src/main/res/xml/file_paths 配置里面需包含 external-path 示例如下
			<?xml version="1.0" encoding="utf-8"?>
			<paths xmlns:android="http://schemas.android.com/apk/res/android">
				<external-path name="external_download" path=""/>
				...
			</paths>

**使用说明：**

	1.提供简单的 UI 弹窗
    //传入约定的json参考Version对应的json
    //remark 更新内容
    //currentLastFlag 是否有新版本  0有新版本 1没有新版本
    //redirectUrl 更新连接
    //forceFlag 是否强制更新 0非强制更新 1强制
    var upgradeJson={
		"forceFlag": 1,
		"currentLastFlag": 0,
		"remark": "我是一个版本更新，此版本是强制更新，我是更新内容",
		"redirectUrl": "https://*******.apk"
    };
    FlutterAppUpgrader.showUpgradeDialog(context, upgradeJson);

    2.提供直接更新的方法
    //传入更新地址，安卓完成下载安装，iOS跳转到AppStore
    //含下载进度/失败回调
    //url 更新的连接 iOS是跳转app store地址 android是apk下载地址
    //downLoadCallBack 下载过程中的回调
    //error 异常信息的回调
    FlutterAppUpgrader.upgradeApp(
    	"https://*****.apk",
        downLoadCallBack: (current, total) {
        	print("共$total,已下载$current");
        },
        error: (err) {
        	print("更新异常：$err");
        },
    );

    3.提供检查IOS AppStore里是否有不同版本的方法
    bool hasNewVersion = await FlutterAppUpgrader.hasNewVersionInAppStore();




CHANGELOG.md
## 0.0.1

* flutter app升级 安卓下载apk后安装 iOS去appStore更新


test/flutter_app_upgrade_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_app_upgrader/flutter_app_upgrader.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_app_upgrade');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await FlutterAppUpgrader.platformVersion, '42');
  });
}


lib/version.dart
class Version {
  Version();

  ///更新内容
  String remark;

  ///设备类型
  String deviceType;

  ///0有新版本 1没有新版本
  num currentLastFlag;

  ///更新连接
  String redirectUrl;

  ///0非强制更新 1强制
  num forceFlag;

  ///创建时间
  num createTime;

  ///版本号
  String version;

  factory Version.fromJson(Map json) {
    return Version()
      ..remark = json['remark']?.toString()
      ..deviceType = json['deviceType']?.toString()
      ..currentLastFlag = json['currentLastFlag'] is num ? json['currentLastFlag'] : null
      ..redirectUrl = json['redirectUrl']?.toString()
      ..forceFlag = json['forceFlag'] is num ? json['forceFlag'] : null
      ..createTime = json['createTime'] is num ? json['createTime'] : null
      ..version = json['version']?.toString();
  }

  Map<String, dynamic> toJson() {
    return {
      'remark': remark,
      'deviceType': deviceType,
      'currentLastFlag': currentLastFlag,
      'redirectUrl': redirectUrl,
      'forceFlag': forceFlag,
      'createTime': createTime,
    };
  }

  bool isValidNewVersion() {
    bool isValid = currentLastFlag == 0 &&
        forceFlag != null &&
        (redirectUrl?.isNotEmpty ?? false) &&
        (redirectUrl.startsWith("http"));
    if (!isValid) {
      print("当前版本不是有效版本，请检查传入Json是否异常！");
    }
    return isValid;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:flutter_app_upgrader/flutter_app_upgrader.dart';
import 'package:permission_handler/permission_handler.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  double _progress;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Plugin example app'),
      ),
      body: Center(
        child: Column(
          children: [
            RaisedButton(
              onPressed: () {
                Map mustJson = {
                  "forceFlag": 1,
                  "currentLastFlag": 0,
                  "remark": "我是一个版本更新，此版本是强制更新",
                  "redirectUrl": "https://gz.bcebos.com/v1/newretail/test/update.apk"
                };
                FlutterAppUpgrader.showUpgradeDialog(context, mustJson);
              },
              child: Text("弹强更窗口"),
            ),
            RaisedButton(
              onPressed: () {
                Map json = {
                  "forceFlag": 0,
                  "currentLastFlag": 0,
                  "remark": "我是一个版本更新，此版本不是强制更新，可跳過",
                  "redirectUrl": "https://gz.bcebos.com/v1/newretail/test/update.apk"
                };
                FlutterAppUpgrader.showUpgradeDialog(context, json);
              },
              child: Text("弹非强更窗口"),
            ),
            RaisedButton(
              onPressed: () {
                Map json = {"forceFlag": 0, "currentLastFlag": 1, "remark": "我是一个版本更新，此版本不是强制更新，可跳過"};
                FlutterAppUpgrader.showUpgradeDialog(context, json);
              },
              child: Text("传入没有新版本的数据"),
            ),
            RaisedButton(
              onPressed: () {
                if ((_progress ?? 0) == 0) {
                  FlutterAppUpgrader.upgradeApp("https://gz.bcebos.com/v1/newretail/test/update.apk",
                      downLoadCallBack: (current, total) {
                    print("downLoad===current:$current,total:$total");
                    setState(() {
                      _progress = current / total;
                    });
                  }, error: (err) {
                    print("downLoadError:$err");
                    setState(() {
                      _progress = 0;
                    });
                  });
                } else if (_progress == 1) {
                  FlutterAppUpgrader.installLocalUnInstalledApk();
                }
              },
              child: Text("直接调用更新api   ${(_progress ?? 0) * 100}%"),
            ),
            RaisedButton(
              onPressed: () async {
                bool hasNewVersion = await FlutterAppUpgrader.hasNewVersionInAppStore();
                print("=====$hasNewVersion");
              },
              child: Text("检查IOS当前版本是否是最新的"),
            ),
          ],
        ),
      ),
    );
  }
}


lib/flutter_app_upgrade.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_app_upgrade/update_dialog_widget.dart';
import 'package:flutter_app_upgrade/version.dart';
import 'package:package_info/package_info.dart';
import 'package:url_launcher/url_launcher.dart';

///
/// flutter应用升级插件
///   安卓：下载应用包+安装
///   IOS：根据传入的url跳转到AppStore
///
/// 1.提供简单的UI弹窗
///   传入约定的json参考Version对应的json
/// 2.提供直接更新的方法
///   含下载进度/失败回调
///
/// 权限说明：
///   IOS无
///   安卓清单文件需要含有以下配置
///   <uses-permission android:name="android.permission.INTERNET" />
///   <uses-permission android:name="android.permission.REQUEST_INSTALL_PACKAGES" />
///
///   7.0之后的provider配置
///     清单文件application结点内添加
///       <provider
///             android:name="androidx.core.content.FileProvider"
///             android:authorities="{包名}.fileprovider"
///             android:exported="false"
///            android:grantUriPermissions="true">
///             <meta-data
///                 android:name="android.support.FILE_PROVIDER_PATHS"
///                 android:resource="@xml/file_paths" />
///        </provider>
///
///     res/xml/file_paths配置里面需包含 external-path 示例如下
///       <?xml version="1.0" encoding="utf-8"?>
///       <paths xmlns:android="http://schemas.android.com/apk/res/android">
///          <external-path name="external_download" path=""/>
///       </paths>
///
///
class FlutterAppUpgrade {
  static const MethodChannel _channel = const MethodChannel('flutter_app_upgrade');

  static Future<String> get platformVersion async {
    final String version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  static Future<bool> _installApk() {
    return _channel.invokeMethod<bool>('installApk');
  }

  static Future<String> _getExternalFilesDirPath() {
    return _channel.invokeMethod<String>('getExternalFilesDirPath');
  }

  static Future<String> _getPackageName() {
    return _channel.invokeMethod<String>('getPackageName');
  }

  ///删除本地已下载的apk包
  ///每次插件引擎挂载成功之后默认会去清除已下载的apk包
  ///极少数特殊情况如有需要也可调用此api来主动清除，比如个人中心设置中的清除缓操作~
  static Future<bool> deleteApk() {
    return _channel.invokeMethod<bool>('deleteApk');
  }

  ///安装本地已经下载好的包
  ///例如如果用户下载完了没有安装又返回到弹窗页面，下载进度100%，再次点击立即更新，则直接安装文件
  static void installLocalUnInstalledApk() async {
    _installApk();
  }

  ///默认提供的ui显示版本更新提示框
  ///versionJson:版本更新json
  ///更新内容remark
  ///是否有新版本 currentLastFlag 0有新版本 1没有新版本
  ///更新连接redirectUrl;
  ///是否强制更新 forceFlag 0非强制更新 1强制
  ///Map mustJson = {"forceFlag": 1, "currentLastFlag": 0, "remark": "我是一个版本更新，此版本是强制更新", "redirectUrl": "http"};
  ///Map json = {"forceFlag": 0, "currentLastFlag": 0, "remark": "我是一个版本更新，此版本不是强制更新，可跳過", "redirectUrl": "http"};
  static void showUpgradeDialog(BuildContext context, Map versionJson) {
    if (context == null || versionJson == null) return;
    Version version = Version.fromJson(versionJson);
    if (version.isValidNewVersion())
      showDialog(
        context: context,
        barrierDismissible: version.forceFlag == 0,
        builder: (context) {
          return UpdateDialogWidget(version);
        },
      );
  }

  ///不使用默认的ui来更新app
  ///安卓下载+安装
  ///iOS去应用市场
  ///url:更新app连接
  ///downLoadCallBack:下载回调
  ///error:更新异常
  static void upgradeApp(String url, {Function downLoadCallBack, Function error}) {
    if (Platform.isIOS) {
      _toAppStore(url, error: error);
    } else if (Platform.isAndroid) {
      _downloadApk(
        url,
        downLoadCallBack: downLoadCallBack,
        error: error,
      );
    }
  }

  static Future<bool> hasNewVersionInAppStore() async {
    if (Platform.isIOS) {
      var info = await PackageInfo.fromPlatform();
      String url = "https://itunes.apple.com/cn/lookup?bundleId=${info.packageName}";
      // String url = "https://itunes.apple.com/cn/lookup?bundleId=cn.com.pateo.shouchebao";
      var response = await Dio().request(url);
      if (response.data is String) {
        var data = JsonDecoder().convert(response.data);
        if (data is Map &&
            data['results'] != null &&
            data['results'] is List &&
            data['results'].length > 0 &&
            data['results'][0] is Map) {
          String version = data['results'][0]['version']?.toString() ?? "";
          return Future.value(version != info.version);
        } else
          return Future.value(false);
      } else
        return Future.value(false);
    } else {
      print("此方法只能应用于iOS平台");
      return Future.value(false);
    }
  }

  static void _toAppStore(String url, {Function error}) async {
    if (await canLaunch(url)) {
      await launch(url);
    } else {
      print("IOS跳转地址异常:$url");
      error?.call("IOS跳转地址异常:$url");
    }
  }

  static void _downloadApk(String url, {Function downLoadCallBack, Function error}) async {
    if (!Platform.isAndroid) {
      print("此方法只能在Android平台使用");
      error?.call("此方法只能在Android平台使用");
      return;
    }
    if (url?.isEmpty ?? true) {
      print("下载地址为空");
      error?.call("下载地址为空");
      return;
    }
    if (!url.startsWith("http")) {
      print("下载地址异常:$url");
      error?.call("下载地址异常:$url");
      return;
    }
    String externalFilesDirPath = await _getExternalFilesDirPath();
    String packageName = await _getPackageName();
    String saveFilePath = '$externalFilesDirPath/$packageName.apk';
    // _installApk();
    // return;
    await Dio()
        .download(
          url,
          saveFilePath,
          onReceiveProgress: downLoadCallBack,
          options: Options(receiveTimeout: 5 * 60 * 1000),
        )
        .then((value) => _installApk())
        .catchError(
      (err) {
        deleteApk();
        error?.call(err);
      },
    );
  }
}


lib/update_dialog_widget.dart
/* 
* created by 1129502088@qq.com
* Date 2021/4/25 10:43
*/

import 'package:flutter/material.dart';
import 'package:flutter_app_upgrader/flutter_app_upgrader.dart';
import 'package:flutter_app_upgrader/version.dart';

class UpdateDialogWidget extends StatefulWidget {
  final Version version;

  UpdateDialogWidget(this.version);

  @override
  _UpdateDialogWidgetState createState() => _UpdateDialogWidgetState();
}

class _UpdateDialogWidgetState extends State<UpdateDialogWidget> {
  double _progress = 0;

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () => Future.value(widget.version?.forceFlag == 0),
      child: UnconstrainedBox(
        constrainedAxis: Axis.vertical,
        child: SizedBox(
          width: MediaQuery.of(context).size.width / 3 * 2,
          child: Dialog(
            insetPadding: EdgeInsets.zero,
            child: Container(
              decoration: BoxDecoration(color: Colors.white, borderRadius: BorderRadius.all(Radius.circular(12))),
              child: Column(
                mainAxisSize: MainAxisSize.min,
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Container(
                    height: 50,
                    child: Center(
                      child: Text(
                        '发现新版本',
                        style: TextStyle(fontWeight: FontWeight.w600, fontSize: 16),
                      ),
                    ),
                  ),
                  Container(
                    margin: EdgeInsets.only(bottom: 10),
                    child: SingleChildScrollView(
                      padding: EdgeInsets.symmetric(horizontal: 20),
                      child: Text(
                        "${widget.version?.remark ?? ""}",
                        style: TextStyle(color: Color(0xFF333333)),
                      ),
                    ),
                    constraints: BoxConstraints(maxHeight: MediaQuery.of(context).size.height / 2 - 50 * 2 - 3),
                  ),
                  Offstage(
                    child: LinearProgressIndicator(
                      value: _progress,
                      minHeight: 3,
                      backgroundColor: Colors.white,
                      valueColor: AlwaysStoppedAnimation<Color>(Colors.blue),
                    ),
                    offstage: _progress == 0,
                  ),
                  Container(
                    decoration: BoxDecoration(
                      border: Border(top: BorderSide(color: Color(0XFFD8D8D8), width: 0.5)),
                    ),
                    height: 40,
                    child: Row(
                      children: [
                        if (widget.version?.forceFlag == 0)
                          Expanded(
                            child: InkWell(
                              onTap: () {
                                if (_progress == 0 || _progress == 1) Navigator.pop(context);
                              },
                              child: Center(
                                child: Text(
                                  '暂不更新',
                                  style: TextStyle(color: Color(0xFF999999)),
                                ),
                              ),
                            ),
                          ),
                        Expanded(
                          child: InkWell(
                            onTap: () {
                              if (_progress == 0) {
                                FlutterAppUpgrader.upgradeApp("${widget.version?.redirectUrl ?? ""}",
                                    downLoadCallBack: (current, total) {
                                  print("downLoad===current:$current,total:$total");
                                  setState(() {
                                    _progress = current / total;
                                  });
                                }, error: (err) {
                                  print("downLoadError:$err");
                                  setState(() {
                                    _progress = 0;
                                  });
                                });
                              } else if (_progress == 1) {
                                FlutterAppUpgrader.installLocalUnInstalledApk();
                              }
                            },
                            child: Center(
                              child: Text(
                                '立即更新',
                                style: TextStyle(fontWeight: FontWeight.w500, fontSize: 14),
                              ),
                            ),
                          ),
                        ),
                      ],
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}


android/src/main/java/com/example/flutter_app_upgrade/FlutterAppUpgradePlugin.java
package com.example.flutter_app_upgrade;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;

import androidx.annotation.NonNull;
import androidx.core.content.FileProvider;

import java.io.File;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * FlutterAppUpgradePlugin
 */
public class FlutterAppUpgradePlugin implements FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;

    private Context context;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        context = flutterPluginBinding.getApplicationContext();
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_app_upgrade");
        channel.setMethodCallHandler(this);
        deleteApk();
    }


    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("getPlatformVersion")) {
            result.success("Android " + android.os.Build.VERSION.RELEASE);
        } else if (call.method.equals("getExternalFilesDirPath")) {
            result.success(context.getExternalFilesDir(null).getAbsolutePath());
        } else if (call.method.equals("getPackageName")) {
            result.success(context.getPackageName());
        } else if (call.method.equals("installApk")) {
            try {
                installApk();
                result.success(true);
            } catch (Exception e) {
                System.out.println(e.getMessage());
                result.success(false);
            }
        } else if (call.method.equals("deleteApk")) {
            try {
                deleteApk();
                result.success(true);
            } catch (Exception e) {
                System.out.println(e.getMessage());
                result.success(false);
            }
        } else {
            result.notImplemented();
        }
    }

    private void deleteApk() {
        String apkPath = context.getExternalFilesDir(null).getAbsolutePath() + "/" + context.getPackageName() + ".apk";
        System.out.println("apkPath:" + apkPath);
        File file = new File(apkPath);
        if (file.isFile() && file.exists()) {
            file.delete();
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }


    private void installApk() {
        String apkPath = context.getExternalFilesDir(null).getAbsolutePath() + "/" + context.getPackageName() + ".apk";
        File apk = new File(apkPath);
//        System.out.println(apkPath + "=====" + apk.isFile() + "===" + apk.exists());
        if (!apk.isFile()) return;
        if (!apk.exists()) return;
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            Uri uri = FileProvider.getUriForFile(context, context.getPackageName() + ".fileprovider", apk);
            intent.setDataAndType(uri, "application/vnd.android.package-archive");
        } else {
            intent.setDataAndType(Uri.fromFile(apk), "application/vnd.android.package-archive");
        }

        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);
    }
}

lib/flutter_app_upgrader.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_app_upgrader/update_dialog_widget.dart';
import 'package:flutter_app_upgrader/version.dart';
import 'package:package_info/package_info.dart';
import 'package:url_launcher/url_launcher.dart';

///
/// flutter应用升级插件
///   安卓：下载应用包+安装
///   IOS：根据传入的url跳转到AppStore
///
/// 1.提供简单的UI弹窗
///   传入约定的json参考Version对应的json
/// 2.提供直接更新的方法
///   含下载进度/失败回调
///
/// 权限说明：
///   IOS无
///   安卓清单文件需要含有以下配置
///   <uses-permission android:name="android.permission.INTERNET" />
///   <uses-permission android:name="android.permission.REQUEST_INSTALL_PACKAGES" />
///
///   7.0之后的provider配置
///     清单文件application结点内添加
///       <provider
///             android:name="androidx.core.content.FileProvider"
///             android:authorities="{包名}.fileprovider"
///             android:exported="false"
///            android:grantUriPermissions="true">
///             <meta-data
///                 android:name="android.support.FILE_PROVIDER_PATHS"
///                 android:resource="@xml/file_paths" />
///        </provider>
///
///     res/xml/file_paths配置里面需包含 external-path 示例如下
///       <?xml version="1.0" encoding="utf-8"?>
///       <paths xmlns:android="http://schemas.android.com/apk/res/android">
///          <external-path name="external_download" path=""/>
///       </paths>
///
///
class FlutterAppUpgrader {
  static const MethodChannel _channel = const MethodChannel('flutter_app_upgrade');

  static Future<String> get platformVersion async {
    final String version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  static Future<bool> _installApk() {
    return _channel.invokeMethod<bool>('installApk');
  }

  static Future<String> _getExternalFilesDirPath() {
    return _channel.invokeMethod<String>('getExternalFilesDirPath');
  }

  static Future<String> _getPackageName() {
    return _channel.invokeMethod<String>('getPackageName');
  }

  ///删除本地已下载的apk包
  ///每次插件引擎挂载成功之后默认会去清除已下载的apk包
  ///极少数特殊情况如有需要也可调用此api来主动清除，比如个人中心设置中的清除缓操作~
  static Future<bool> deleteApk() {
    return _channel.invokeMethod<bool>('deleteApk');
  }

  ///安装本地已经下载好的包
  ///例如如果用户下载完了没有安装又返回到弹窗页面，下载进度100%，再次点击立即更新，则直接安装文件
  static void installLocalUnInstalledApk() async {
    _installApk();
  }

  ///默认提供的ui显示版本更新提示框
  ///versionJson:版本更新json
  ///更新内容remark
  ///是否有新版本 currentLastFlag 0有新版本 1没有新版本
  ///更新连接redirectUrl;
  ///是否强制更新 forceFlag 0非强制更新 1强制
  ///Map mustJson = {"forceFlag": 1, "currentLastFlag": 0, "remark": "我是一个版本更新，此版本是强制更新", "redirectUrl": "http"};
  ///Map json = {"forceFlag": 0, "currentLastFlag": 0, "remark": "我是一个版本更新，此版本不是强制更新，可跳過", "redirectUrl": "http"};
  static void showUpgradeDialog(BuildContext context, Map versionJson) {
    if (context == null || versionJson == null) return;
    Version version = Version.fromJson(versionJson);
    if (version.isValidNewVersion())
      showDialog(
        context: context,
        barrierDismissible: version.forceFlag == 0,
        builder: (context) {
          return UpdateDialogWidget(version);
        },
      );
  }

  ///不使用默认的ui来更新app
  ///安卓下载+安装
  ///iOS去应用市场
  ///url:更新app连接
  ///downLoadCallBack:下载回调
  ///error:更新异常
  static void upgradeApp(String url, {Function downLoadCallBack, Function error}) {
    if (Platform.isIOS) {
      _toAppStore(url, error: error);
    } else if (Platform.isAndroid) {
      _downloadApk(
        url,
        downLoadCallBack: downLoadCallBack,
        error: error,
      );
    }
  }

  static Future<bool> hasNewVersionInAppStore() async {
    if (Platform.isIOS) {
      var info = await PackageInfo.fromPlatform();
      String url = "https://itunes.apple.com/cn/lookup?bundleId=${info.packageName}";
      // String url = "https://itunes.apple.com/cn/lookup?bundleId=cn.com.pateo.shouchebao";
      var response = await Dio().request(url);
      if (response.data is String) {
        var data = JsonDecoder().convert(response.data);
        if (data is Map &&
            data['results'] != null &&
            data['results'] is List &&
            data['results'].length > 0 &&
            data['results'][0] is Map) {
          String version = data['results'][0]['version']?.toString() ?? "";
          return Future.value(version != info.version);
        } else
          return Future.value(false);
      } else
        return Future.value(false);
    } else {
      print("此方法只能应用于iOS平台");
      return Future.value(false);
    }
  }

  static void _toAppStore(String url, {Function error}) async {
    if (await canLaunch(url)) {
      await launch(url);
    } else {
      print("IOS跳转地址异常:$url");
      error?.call("IOS跳转地址异常:$url");
    }
  }

  static void _downloadApk(String url, {Function downLoadCallBack, Function error}) async {
    if (!Platform.isAndroid) {
      print("此方法只能在Android平台使用");
      error?.call("此方法只能在Android平台使用");
      return;
    }
    if (url?.isEmpty ?? true) {
      print("下载地址为空");
      error?.call("下载地址为空");
      return;
    }
    if (!url.startsWith("http")) {
      print("下载地址异常:$url");
      error?.call("下载地址异常:$url");
      return;
    }
    String externalFilesDirPath = await _getExternalFilesDirPath();
    String packageName = await _getPackageName();
    String saveFilePath = '$externalFilesDirPath/$packageName.apk';
    // _installApk();
    // return;
    await Dio()
        .download(
          url,
          saveFilePath,
          onReceiveProgress: downLoadCallBack,
          options: Options(receiveTimeout: 5 * 60 * 1000),
        )
        .then((value) => _installApk())
        .catchError(
      (err) {
        deleteApk();
        error?.call(err);
      },
    );
  }
}
